architecture/2025-06-graph-db-replace-planes.md

```diff
@@ -0,0 +1,616 @@
+    ```json
+    {
+      "gateway": "api-gateway",
+      "connected_secrets": [
+        {
+          "name": "tls-cert",
+          "type": "Applications.Core/secret",
+          "properties": {
+            "type": "certificate",
+            "domain": "wrong-domain.com",
+            "expires": "2024-12-31T23:59:59Z"
+          }
+        },
+        {
+          "name": "tls-key", 
+          "type": "Applications.Core/secret",
+          "properties": {
+            "type": "private-key",
+            "domain": "wrong-domain.com"
+          }
+        }
+      ]
+    }
+    ```
+
+* **User Benefit:** User immediately sees all certificate secrets connected to their gateway and discovers the wrong domain configuration, enabling quick troubleshooting without manually checking each secret individually.
+
+### Design
+
+#### High-Level Design
+
+1.  **Introduce Graph Access Layer (GAL):** The GAL will be integrated as an internal service within Radius components responsible for managing the application graph, initially working alongside existing storage mechanisms with a path toward full database consolidation.
+2.  **PostgreSQL + AGE Container:** PostgreSQL with Apache AGE will be deployed as a separate container in the Radius control plane, providing both traditional database capabilities and graph functionality through Cypher queries.
+3.  **Unified Database Architecture:** This approach establishes PostgreSQL as the foundation for both graph operations and potential future migration of all Radius data storage, simplifying the overall architecture.
+4.  **Schema Definition:** A formal schema for application graph entities (Applications, Environments, Resources as nodes) and their relationships (as edges with types and properties) will be defined and enforced in PostgreSQL + AGE.
+5.  **Component Updates:** Radius components that currently perform application graph operations (traversals, relationship queries) will be updated to use the GAL for graph queries while maintaining existing functionality through current storage mechanisms during the transition period.
+
+**Architecture Rationale:** PostgreSQL with Apache AGE provides a strategic foundation that can serve both immediate graph database needs and future storage consolidation. Unlike embedded solutions, a containerized PostgreSQL deployment offers: (1) Production-ready operational patterns familiar to most teams; (2) Native support for both relational and graph data models; (3) Established ecosystem of monitoring, backup, and scaling tools; (4) The flexibility to migrate additional Radius storage needs to the same technology stack over time, reducing operational complexity. **Note: This proposal is scoped specifically to graph operations via the GAL. Future work would involve migrating key-value operations from etcd to PostgreSQL and renaming the GAL to DAL (Data Access Layer) to reflect its expanded role as the unified data access interface for all Radius storage operations.**
```

Comment:
> **Note: This proposal is scoped specifically to graph operations via the GAL. Future work would involve migrating key-value operations from etcd to PostgreSQL and renaming the GAL to DAL (Data Access Layer) to reflect its expanded role as the unified data access interface for all Radius storage operations.**

Will data be duplicated across both etcd and the graph database in the meantime?